contributors/design-proposals/storage/csi-migration.md

```diff
@@ -224,78 +284,300 @@ the same ways as we are not removing/adding PV’s or PVC’s in out of the ordi
+	// * "csi-plugin": the attach/detach controller ought to request
+	//   the csi plugin to perform the volume operation rather than perform it directly.
+	VolumePluginMechanism VolumePluginMechanism `json:"volumePluginMechanism"`
+}
+
+// CSIDriverInfoSpec holds information about the specification of one CSI driver installed on a node
+type CSIDriverInfoSpec struct {
+	// name is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Name string `json:"name"`
+
+	// nodeID of the node from the driver point of view.
+	// This field enables Kubernetes to communicate with storage systems that do
+	// not share the same nomenclature for nodes. For example, Kubernetes may
+	// refer to a given node as "node1", but the storage system may refer to
+	// the same node as "nodeA". When Kubernetes issues a command to the storage
+	// system to attach a volume to a specific node, it can use this field to
+	// refer to the node name using the ID that the storage system will
+	// understand, e.g. "nodeA" instead of "node1".
+	NodeID string `json:"nodeID"`
+
+	// topologyKeys is the list of keys supported by the driver.
+	// When a driver is initialized on a cluster, it provides a set of topology
+	// keys that it understands (e.g. "company.com/zone", "company.com/region").
+	// When a driver is initialized on a node it provides the same topology keys
+	// along with values that kubelet applies to the coresponding node API
+	// object as labels.
+	// When Kubernetes does topology aware provisioning, it can use this list to
+	// determine which labels it should retrieve from the node object and pass
+	// back to the driver.
+	TopologyKeys []string `json:"topologyKeys"`
+}
+```
+
+#### Spec/Status Split
+
+The new `CSINodeInfo` API has been split up to have a `Spec` and a `Status`. The
+reason being that there are some fields that are constant after being set by the
+CSI Driver on initialization: `NodeID` and `TopologyKeys`. Whereas there are
+some other fields that are subject to change: `Available` and
+`VolumePluginMechanism`. There are future plans to reconcile the `Status` based
+on periodically polling the driver.
+
+Because of this split there is a possibility that there will be a `Spec` with no
+corresponding `Status` or a `Status` without a corresponding `Spec`. All
+components reacting to the `CSINodeInfo` API must be resilient to this
+possibility. It should be assumed that the driver is not available when a `Spec`
+without a `Status` is observed. When a `Status` without a `Spec` is observed we
+are in an undefined state and cannot assume anything about the CSI Driver.
+
+#### New Fields
+
+The important fields to note are `Available` which allows the Attach/Detach
+controller to tell whether the driver is available on each node or not. As well
+as `VolumePluginMechanism` which is set by the Kubelet depending on the feature
```

Comment:
Is `VolumePluginMechanism` a temporary field that is no longer needed once migration is GA?  If so, maybe we should make it an annotation instead of a field if we plan to remove it eventually.

Reply:
I wonder if making it an annotation would solve some of the strange lifecycle behaviors with the API